packages/frontend/native/media_capture/src/macos/tap_audio.rs

```diff
@@ -189,30 +221,246 @@ impl AggregateDevice {
+    };
+
+    // Configure the aggregate device to ensure proper handling of both input and
+    // output
+    device.configure_aggregate_device()?;
+
+    // Activate both the input and output devices and store their proc IDs
+    let input_proc_id = device.activate_audio_device(input_device_id)?;
+    let output_proc_id = device.activate_audio_device(output_device_id)?;
+
+    device.input_proc_id = Some(input_proc_id);
+    device.output_proc_id = Some(output_proc_id);
+
+    Ok(device)
   }
 
-  pub fn start(
-    &mut self,
-    audio_stream_callback: Arc<ThreadsafeFunction<Float32Array, (), Float32Array, true>>,
-  ) -> Result<AudioTapStream> {
-    // Read and log the audio format before starting the device
+  // Configures the aggregate device to ensure proper handling of both input and
+  // output streams
+  fn configure_aggregate_device(&self) -> Result<AudioStats> {
+    // Read the current audio format to ensure it's properly configured
+    let audio_format = read_audio_stream_basic_description(self.tap_id)?;
+
+    // Create initial audio stats with the actual sample rate but always use mono
+    let initial_sample_rate = audio_format.0.mSampleRate;
     let mut audio_stats = AudioStats {
-      sample_rate: 44100.0,
+      sample_rate: initial_sample_rate,
       channels: 1, // Always set to 1 channel (mono)
     };
 
-    if let Ok(audio_format) = read_audio_stream_basic_description(self.tap_id) {
-      // Store the audio format information
-      audio_stats.sample_rate = audio_format.0.mSampleRate;
-      // Always use 1 channel regardless of what the system reports
-      audio_stats.channels = 1;
+    // Set the preferred sample rate on the device
+    // This is similar to how Screen Capture Kit allows setting the sample rate
+    let preferred_sample_rate = initial_sample_rate; // Use the device's current sample rate
+
+    // First, check if the preferred sample rate is available
+    let mut is_sample_rate_available = false;
+    let mut best_available_rate = preferred_sample_rate; // Default to preferred rate
+
+    unsafe {
+      // Get the available sample rates
+      let address = AudioObjectPropertyAddress {
+        mSelector: kAudioDevicePropertyAvailableNominalSampleRates,
+        mScope: kAudioObjectPropertyScopeGlobal,
+        mElement: kAudioObjectPropertyElementMain,
+      };
+
+      // Get the size of the property data
+      let mut data_size: u32 = 0;
+      let status = AudioObjectGetPropertyDataSize(
+        self.id,
+        &address as *const AudioObjectPropertyAddress,
+        0,
+        std::ptr::null(),
+        &mut data_size as *mut u32,
+      );
+
+      if status == 0 && data_size > 0 {
+        // Calculate how many ranges we have
+        let range_count = data_size as usize / std::mem::size_of::<AudioValueRange>();
+
+        // Allocate memory for the ranges
+        let mut ranges: Vec<AudioValueRange> = Vec::with_capacity(range_count);
+        ranges.set_len(range_count);
```

Comment:
`with_capacity` will alloc the required mems, `set_len` is unsafe and will make it undefined behavior if   visited within the range

Reply:
I am not sure, maybe this is some limitation when dealing with obj-c api?